actor/builtin/miner/miner.go

```diff
@@ -955,6 +973,70 @@ func (ma *Actor) SubmitPoSt(ctx exec.VMContext, poStProofs []types.PoStProof, do
 	return 0, nil
 }
 
+// SlashStorageFault is called by an independent actor to remove power and
+// take collateral from this miner when the miner has failed to submit a
+// PoSt on time.
+func (ma *Actor) SlashStorageFault(ctx exec.VMContext) (uint8, error) {
+	if err := ctx.Charge(actor.DefaultGasCost); err != nil {
+		return exec.ErrInsufficientGas, errors.RevertErrorWrap(err, "Insufficient gas")
+	}
+
+	chainHeight := ctx.BlockHeight()
+	var state State
+	_, err := actor.WithState(ctx, &state, func() (interface{}, error) {
+		// You can only be slashed once for missing your PoSt.
+		if state.SlashedAt != nil {
+			return nil, errors.NewCodedRevertError(ErrMinerAlreadySlashed, "miner already slashed")
+		}
+
+		// Only a miner who is expected to prove, can be slashed.
+		if state.ProvingSet.Size() == 0 {
+			return nil, errors.NewCodedRevertError(ErrMinerNotSlashable, "miner is inactive")
+		}
+
+		// Only if the miner is actually late, they can be slashed.
+		deadline := state.ProvingPeriodEnd.Add(GenerationAttackTime(state.SectorSize))
+		if chainHeight.LessEqual(deadline) {
+			return nil, errors.NewCodedRevertError(ErrMinerNotSlashable, "miner not yet tardy")
+		}
+
+		// Strip the miner of their power.
+		powerDelta := types.ZeroBytes.Sub(state.Power) // negate bytes amount
+		_, ret, err := ctx.Send(address.StorageMarketAddress, "updateStorage", types.ZeroAttoFIL, []interface{}{powerDelta})
+		if err != nil {
+			return nil, err
+		}
+		if ret != 0 {
+			return nil, Errors[ErrStoragemarketCallFailed]
+		}
+		state.Power = types.NewBytesAmount(0)
+
+		// record what has been slashed
+		state.SlashedSet = state.ProvingSet
+
+		// reserve collateral for arbitration
+		// TODO: This calculation is probably not correct: https://github.com/filecoin-project/go-filecoin/issues/3050
+		state.OwedStorageCollateral = ma.getPledgeCollateralRequirement(state, chainHeight)
```

Comment:
This is definitely wrong. I think we should use zero here until a workable definition emerges: https://github.com/filecoin-project/specs/issues/385, https://github.com/filecoin-project/specs/issues/386